ampd/src/mvx/mod.rs

```diff
@@ -24,15 +25,20 @@ pub struct WeightedSigners {
 
 impl WeightedSigners {
     pub fn hash(&self) -> Hash {
-        let mut encoded = Vec::new();
+        let mut encoded: Vec<Vec<u8>> = Vec::new();

```

Comment:
Do you need `Vec<Vec<u8>>`? `digest` takes a `Ref<[u8]>`. If possible I'd prefer to use `Vec<&[u8]>` as before. Defining encoded as mutable with nested vectors allows you to modify the vector elements as well I believe, which seems dangerous

Reply:
The reason I didn't use `Vec<&[u8]>` is because of the needed call to `usize_to_u32` which is done in the `for` loop. That creates a new owned variable that doesn't live long enough, hence I needed to change this to to a `Vec<Vec<u8>>`.

The solution would have been to save the length in the `WeightedSigner` & `WeightedSigners` structs so I can reference it, but I don't think that is a good idea.

Reply:
but now that issue doesn't exist anymore, does it? You can just call `encoded.push(&usize_to_u32(self.signers.len()));

Reply:
Still not possible since the value will live only long enough for the `for` loop. This code does not work because the `len` from the for doesn't live long enough. I tried multiple ways of refactoring to storing refs, but I didn't find any solution:
```rust
pub fn hash(&self) -> Hash {
        let mut encoded: Vec<&[u8]> = Vec::new();

        let len = usize_to_u32(self.signers.len());
        encoded.push(&len);

        for signer in self.signers.iter() {
            encoded.push(&signer.signer);

            let len = usize_to_u32(signer.weight.len());
            encoded.push(&len);
            encoded.push(&signer.weight);
        }

        let len = usize_to_u32(self.threshold.len());
        encoded.push(&len);
        encoded.push(&self.threshold);
        encoded.push(&self.nonce);

        Keccak256::digest(encoded.concat()).into()
    }
```

```
let len = usize_to_u32(signer.weight.len());
   |                 --- binding `len` declared here
37 |             encoded.push(&len);
   |                          ^^^^ borrowed value does not live long enough
```